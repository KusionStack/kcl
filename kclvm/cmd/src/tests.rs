use crate::{app, settings::build_settings};

#[test]
fn test_build_settings() {
    let work_dir = work_dir();
    let matches = app().get_matches_from(settings_arguments(work_dir.join("kcl.yaml")));
    let matches = matches.subcommand_matches("run").unwrap();
    let s = build_settings(matches).unwrap();
    // Testing work directory
    assert_eq!(s.path().as_ref().unwrap().to_str(), work_dir.to_str());
    // Testing CLI configs
    assert_eq!(
        s.settings().kcl_cli_configs.as_ref().unwrap().files,
        Some(vec!["hello.k".to_string()])
    );
    assert_eq!(
        s.settings().kcl_cli_configs.as_ref().unwrap().disable_none,
        Some(true)
    );
    assert_eq!(
        s.settings()
            .kcl_cli_configs
            .as_ref()
            .unwrap()
            .strict_range_check,
        Some(true)
    );
    assert_eq!(
        s.settings().kcl_cli_configs.as_ref().unwrap().overrides,
        Some(vec!["c.a=1".to_string(), "c.b=1".to_string(),])
    );
    assert_eq!(
        s.settings().kcl_cli_configs.as_ref().unwrap().path_selector,
        Some(vec!["a.b.c".to_string()])
    );
    assert_eq!(s.settings().input(), vec!["hello.k".to_string()]);
}

#[test]
fn test_build_settings_fail() {
    let matches = app().get_matches_from(settings_arguments(work_dir().join("error_kcl.yaml")));
    let matches = matches.subcommand_matches("run").unwrap();
    assert!(build_settings(matches).is_err());
}

fn work_dir() -> std::path::PathBuf {
    std::path::Path::new(".")
        .join("src")
        .join("test_data")
        .join("settings")
}

fn settings_arguments(path: std::path::PathBuf) -> Vec<String> {
    vec![
        "kclvm_cli".to_string(),
        "run".to_string(),
        "-Y".to_string(),
        path.to_str().unwrap().to_string(),
        "-r".to_string(),
        "-O".to_string(),
        "c.a=1".to_string(),
        "-O".to_string(),
        "c.b=1".to_string(),
        "-S".to_string(),
        "a.b.c".to_string(),
    ]
}
