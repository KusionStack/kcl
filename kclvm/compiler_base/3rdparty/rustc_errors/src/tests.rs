mod test_styled_buffer {
    use compiler_base_style::{option_box_style, diagnostic_style::DiagnosticStyle, Style};

    use crate::styled_buffer::{StyledBuffer, StyledString};


    fn construct_new_styledbuffer() -> StyledBuffer {
        StyledBuffer::new()
    }

    fn putc_hello_world(sb: &mut StyledBuffer) {
        sb.putc(0, 0, 'H',option_box_style!(DiagnosticStyle::NoStyle));
        sb.putc(0, 1, 'e',option_box_style!(DiagnosticStyle::NoStyle));
        sb.putc(0, 2, 'l',option_box_style!(DiagnosticStyle::NoStyle));
        sb.putc(0, 3, 'l',option_box_style!(DiagnosticStyle::NoStyle));
        sb.putc(0, 4, 'o',option_box_style!(DiagnosticStyle::NoStyle));
        sb.putc(0, 5, 'W',option_box_style!(DiagnosticStyle::NeedFix));
        sb.putc(0, 6, 'o',option_box_style!(DiagnosticStyle::NeedFix));
        sb.putc(0, 7, 'r',option_box_style!(DiagnosticStyle::NeedFix));
        sb.putc(0, 8, 'l',option_box_style!(DiagnosticStyle::NeedFix));
        sb.putc(0, 9, 'd',option_box_style!(DiagnosticStyle::NeedFix));
    }

    fn puts_hello_world(sb: &mut StyledBuffer) {
        sb.puts(0, 0, "Hello",option_box_style!(DiagnosticStyle::NoStyle));
        sb.puts(0, 5, "World",option_box_style!(DiagnosticStyle::NeedFix));
    }

    fn putl_hello_world(sb: &mut StyledBuffer) {
        sb.putl("Hello",option_box_style!(DiagnosticStyle::NoStyle));
        sb.putl("World",option_box_style!(DiagnosticStyle::NeedFix));
    }

    fn appendl_hello_world(sb: &mut StyledBuffer) {
        sb.appendl("Hello",option_box_style!(DiagnosticStyle::NoStyle));
        sb.appendl("World",option_box_style!(DiagnosticStyle::NeedFix));
    }

    fn require_hello_world(styled_strings: Vec<Vec<StyledString>>) {
        assert_eq!(styled_strings.len(), 1);
        assert_eq!(styled_strings.get(0).unwrap().len(), 2);

        
        assert_eq!(styled_strings.get(0).unwrap().get(0).unwrap().text, "Hello");
        assert!(DiagnosticStyle::NoStyle.style_eq(styled_strings.get(0).unwrap().get(0).unwrap().style.as_ref().unwrap()));
        assert_eq!(styled_strings.get(0).unwrap().get(1).unwrap().text, "World");
        assert!(DiagnosticStyle::NeedFix.style_eq(styled_strings.get(0).unwrap().get(1).unwrap().style.as_ref().unwrap()));
    }

    #[test]
    fn test_putc() {
        let mut sb = construct_new_styledbuffer();
        putc_hello_world(&mut sb);

        let styled_strings = sb.render();

        require_hello_world(styled_strings);

        sb.putc(0, 0, 'H',option_box_style!(DiagnosticStyle::NoStyle));
        sb.putc(0, 1, 'E',option_box_style!(DiagnosticStyle::NoStyle));
        sb.putc(0, 2, 'L',option_box_style!(DiagnosticStyle::NoStyle));
        sb.putc(0, 3, 'L',option_box_style!(DiagnosticStyle::NoStyle));
        sb.putc(0, 4, 'O',option_box_style!(DiagnosticStyle::NoStyle));
        let styled_strings = sb.render();
        assert_eq!(styled_strings.get(0).unwrap().get(0).unwrap().text, "HELLO");
        assert!(DiagnosticStyle::NoStyle.style_eq(styled_strings.get(0).unwrap().get(0).unwrap().style.as_ref().unwrap()),
        "style error: expected style : {:?}", DiagnosticStyle::NoStyle);
    }

    #[test]
    fn test_putc_new_line() {
        let mut sb = construct_new_styledbuffer();
        putc_hello_world(&mut sb);

        sb.putc(2, 0, 'A',option_box_style!(DiagnosticStyle::Important));
        let styled_strings = sb.render();
        assert_eq!(styled_strings.len(), 3);
        assert_eq!(styled_strings.get(0).unwrap().len(), 2);
        assert_eq!(styled_strings.get(1).unwrap().len(), 0);
        assert_eq!(styled_strings.get(2).unwrap().len(), 1);
        assert_eq!(styled_strings.get(2).unwrap().get(0).unwrap().text, "A");
        assert!(DiagnosticStyle::Important.style_eq(styled_strings.get(2).unwrap().get(0).unwrap().style.as_ref().unwrap()),
        "style error: expected style : {:?}", DiagnosticStyle::Important);
    }

    #[test]
    fn test_puts() {
        let mut sb = construct_new_styledbuffer();
        puts_hello_world(&mut sb);
        let styled_strings = sb.render();
        require_hello_world(styled_strings);
    }

    #[test]
    fn test_puts_new_line() {
        let mut sb = construct_new_styledbuffer();
        puts_hello_world(&mut sb);

        sb.puts(2, 0, "A",option_box_style!(DiagnosticStyle::Important));
        let styled_strings = sb.render();
        assert_eq!(styled_strings.len(), 3);
        assert_eq!(styled_strings.get(0).unwrap().len(), 2);
        assert_eq!(styled_strings.get(1).unwrap().len(), 0);
        assert_eq!(styled_strings.get(2).unwrap().len(), 1);
        assert_eq!(styled_strings.get(2).unwrap().get(0).unwrap().text, "A");
        assert!(DiagnosticStyle::Important.style_eq(styled_strings.get(2).unwrap().get(0).unwrap().style.as_ref().unwrap()),
        "style error: expected style : {:?}", DiagnosticStyle::Important);
    }

    #[test]
    fn test_putl() {
        let mut sb = construct_new_styledbuffer();
        putl_hello_world(&mut sb);
        let styled_strings = sb.render();
        assert_eq!(styled_strings.len(), 2);
        assert_eq!(styled_strings.get(0).unwrap().len(), 1);

        assert_eq!(styled_strings.get(0).unwrap().get(0).unwrap().text, "Hello");
        assert!(DiagnosticStyle::NoStyle.style_eq(styled_strings.get(0).unwrap().get(0).unwrap().style.as_ref().unwrap()),
        "style error: expected style : {:?}", DiagnosticStyle::NoStyle);

        assert_eq!(styled_strings.get(1).unwrap().get(0).unwrap().text, "World");
        assert!(DiagnosticStyle::NeedFix.style_eq(styled_strings.get(1).unwrap().get(0).unwrap().style.as_ref().unwrap()),
        "style error: expected style : {:?}", DiagnosticStyle::NeedFix);
    }

    #[test]
    fn test_appendl() {
        let mut sb = construct_new_styledbuffer();
        appendl_hello_world(&mut sb);
        let styled_strings = sb.render();
        require_hello_world(styled_strings);
    }

    #[test]
    fn test_prepend() {
        let mut sb = construct_new_styledbuffer();
        sb.appendl("World",option_box_style!(DiagnosticStyle::NeedFix));
        sb.prepend(0, "Hello",option_box_style!(DiagnosticStyle::NoStyle));
        let styled_strings = sb.render();
        require_hello_world(styled_strings);
    }

    #[test]
    fn test_num_lines() {
        let mut sb = construct_new_styledbuffer();
        putc_hello_world(&mut sb);
        assert_eq!(sb.num_lines(), 1);
        sb.appendl("World",option_box_style!(DiagnosticStyle::NeedFix));
        assert_eq!(sb.num_lines(), 1);
        putl_hello_world(&mut sb);
        assert_eq!(sb.num_lines(), 3);
        puts_hello_world(&mut sb);
        assert_eq!(sb.num_lines(), 3);
    }
}


