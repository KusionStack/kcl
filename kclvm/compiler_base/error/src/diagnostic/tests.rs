mod test_diagnostic {
    use crate::diagnostic::{components::Label, style::DiagnosticStyle, Component, Diagnostic};
    use rustc_errors::styled_buffer::StyledBuffer;

    #[test]
    fn test_diagnostic_with_label() {
        let mut diagnostic = Diagnostic::new();

        let err_label = Box::new(Label::Error("E3033".to_string()));
        diagnostic.append_component(err_label);

        let msg = Box::new(": this is an error!".to_string());
        diagnostic.append_component(msg);

        let mut sb = StyledBuffer::<DiagnosticStyle>::new();

        let mut errs = vec![];
        diagnostic.format(&mut sb, &mut errs);
        let result = sb.render();

        assert_eq!(result.len(), 1);
        assert_eq!(result.get(0).unwrap().len(), 3);
        assert_eq!(result.get(0).unwrap().get(0).unwrap().text, "error");
        assert_eq!(result.get(0).unwrap().get(1).unwrap().text, "[E3033]");
        assert_eq!(
            result.get(0).unwrap().get(2).unwrap().text,
            ": this is an error!"
        );

        assert_eq!(
            result.get(0).unwrap().get(0).unwrap().style,
            Some(DiagnosticStyle::NeedFix)
        );
        assert_eq!(
            result.get(0).unwrap().get(1).unwrap().style,
            Some(DiagnosticStyle::Helpful)
        );
        assert_eq!(result.get(0).unwrap().get(2).unwrap().style, None);
    }
}

mod test_components {

    use crate::{
        components::StringWithStyle,
        diagnostic::{components::Label, style::DiagnosticStyle, Component},
    };
    use rustc_errors::styled_buffer::StyledBuffer;

    #[test]
    fn test_label() {
        let mut sb = StyledBuffer::<DiagnosticStyle>::new();
        let mut errs = vec![];
        Label::Error("E3030".to_string()).format(&mut sb, &mut errs);
        Label::Warning("W3030".to_string()).format(&mut sb, &mut errs);
        Label::Note.format(&mut sb, &mut errs);
        Label::Help.format(&mut sb, &mut errs);
        let result = sb.render();
        assert_eq!(errs.len(), 0);
        assert_eq!(result.len(), 1);
        assert_eq!(result.get(0).unwrap().len(), 6);
        assert_eq!(result.get(0).unwrap().get(0).unwrap().text, "error");
        assert_eq!(result.get(0).unwrap().get(1).unwrap().text, "[E3030]");
        assert_eq!(result.get(0).unwrap().get(2).unwrap().text, "warning");
        assert_eq!(result.get(0).unwrap().get(3).unwrap().text, "[W3030]");
        assert_eq!(result.get(0).unwrap().get(4).unwrap().text, "note");
        assert_eq!(result.get(0).unwrap().get(5).unwrap().text, "help");
    }

    #[test]
    fn test_string() {
        let mut sb = StyledBuffer::<DiagnosticStyle>::new();
        let mut errs = vec![];
        "this is a component string"
            .to_string()
            .format(&mut sb, &mut errs);
        let result = sb.render();
        assert_eq!(errs.len(), 0);
        assert_eq!(result.len(), 1);
        assert_eq!(result.get(0).unwrap().len(), 1);
        assert_eq!(
            result.get(0).unwrap().get(0).unwrap().text,
            "this is a component string"
        );
        assert_eq!(result.get(0).unwrap().get(0).unwrap().style, None);
    }

    #[test]
    fn test_string_with_style() {
        let mut sb = StyledBuffer::<DiagnosticStyle>::new();
        let mut errs = vec![];
        StringWithStyle::new_with_style(
            "This is a string with NeedFix style".to_string(),
            Some(DiagnosticStyle::NeedFix),
        )
        .format(&mut sb, &mut errs);
        let result = sb.render();
        assert_eq!(errs.len(), 0);
        assert_eq!(result.len(), 1);
        assert_eq!(result.get(0).unwrap().len(), 1);
        assert_eq!(
            result.get(0).unwrap().get(0).unwrap().text,
            "This is a string with NeedFix style"
        );
        assert_eq!(
            result.get(0).unwrap().get(0).unwrap().style.unwrap(),
            DiagnosticStyle::NeedFix
        );

        StringWithStyle::new_with_no_style("This is a string with no style".to_string())
            .format(&mut sb, &mut errs);
        let result = sb.render();
        assert_eq!(errs.len(), 0);
        assert_eq!(result.len(), 1);
        assert_eq!(result.get(0).unwrap().len(), 2);
        assert_eq!(
            result.get(0).unwrap().get(1).unwrap().text,
            "This is a string with no style"
        );
        assert_eq!(result.get(0).unwrap().get(1).unwrap().style, None);
    }
}

mod test_error_message {
    use crate::{diagnostic::diagnostic_message::TemplateLoader, diagnostic_handler::MessageArgs};

    #[test]
    fn test_template_message() {
        let template_dir = "./src/diagnostic/locales/en-US";
        let template_loader = TemplateLoader::new_with_template_dir(template_dir).unwrap();

        let mut args = MessageArgs::new();
        check_template_msg(
            "invalid-syntax",
            None,
            &args,
            "Invalid syntax",
            &template_loader,
        );

        args.set("expected_items", "I am an expected item");
        check_template_msg(
            "invalid-syntax",
            Some("expected"),
            &args,
            "Expected one of `\u{2068}I am an expected item\u{2069}`",
            &template_loader,
        );

        args.set("expected_items", "I am an expected item");
        check_template_msg(
            "invalid-syntax-1",
            Some("expected_1"),
            &args,
            "Expected one of `\u{2068}I am an expected item\u{2069}` 1",
            &template_loader,
        );
    }

    fn check_template_msg(
        index: &str,
        sub_index: Option<&str>,
        args: &MessageArgs,
        expected_msg: &str,
        template_loader: &TemplateLoader,
    ) {
        let msg_in_line = template_loader.get_msg_to_str(index, sub_index, &args);
        assert_eq!(msg_in_line.unwrap(), expected_msg);
    }
}
