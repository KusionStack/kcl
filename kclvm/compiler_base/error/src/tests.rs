mod test_diagnostic_handler {
    use std::panic;

    use crate::{Diagnostic, DiagnosticHandler, DiagnosticStyle, MessageArgs};

    #[test]
    fn test_diagnostic_handler_new_with_template_dir() {
        let diag_handler =
            DiagnosticHandler::new_with_template_dir("./src/diagnostic/locales/en-US/");
        match diag_handler {
            Ok(_) => {}
            Err(_) => {
                panic!("`diag_handler` should be Ok(...)")
            }
        }

        let diag_handler_invalid = DiagnosticHandler::new_with_template_dir("./invalid_path");
        match diag_handler_invalid {
            Ok(_) => {
                panic!("`diag_handler_invalid` should be Err(...)")
            }
            Err(_) => {}
        }
    }

    #[test]
    fn test_diagnostic_handler_add_diagnostic() {
        let diag_1 = Diagnostic::<DiagnosticStyle>::new();
        let diag_handler =
            DiagnosticHandler::new_with_template_dir("./src/diagnostic/locales/en-US/").unwrap();
        assert_eq!(diag_handler.diagnostics_count().unwrap(), 0);

        diag_handler.add_err_diagnostic(diag_1);
        assert_eq!(diag_handler.diagnostics_count().unwrap(), 1);
    }

    #[test]
    fn test_diagnostic_handler_get_diagnostic_msg() {
        let no_args = MessageArgs::new();
        let index = "invalid-syntax";
        let sub_index = None;
        let diag_handler =
            DiagnosticHandler::new_with_template_dir("./src/diagnostic/locales/en-US/").unwrap();
        let msg_in_line_1 = diag_handler
            .get_diagnostic_msg(index, sub_index, &no_args)
            .unwrap();
        assert_eq!(msg_in_line_1, "Invalid syntax");

        let mut args = MessageArgs::new();
        args.set("expected_items", "I am an expected item");
        let sub_index = "expected";
        let msg_in_line_2 = diag_handler
            .get_diagnostic_msg(index, Some(sub_index), &args)
            .unwrap();
        assert_eq!(
            msg_in_line_2,
            "Expected one of `\u{2068}I am an expected item\u{2069}`"
        );
    }

    #[test]
    fn test_diagnostic_handler_has() {
        let diag_handler =
            DiagnosticHandler::new_with_template_dir("./src/diagnostic/locales/en-US/").unwrap();
        // test has_errors()
        assert_eq!(diag_handler.has_errors().unwrap(), false);
        diag_handler.add_err_diagnostic(Diagnostic::<DiagnosticStyle>::new());
        assert_eq!(diag_handler.has_errors().unwrap(), true);

        // test has_warns()
        assert_eq!(diag_handler.has_warns().unwrap(), false);
        diag_handler.add_warn_diagnostic(Diagnostic::<DiagnosticStyle>::new());
        assert_eq!(diag_handler.has_warns().unwrap(), true);
    }

    #[test]
    fn test_abort_if_errors() {
        let diag_handler =
            DiagnosticHandler::new_with_template_dir("./src/diagnostic/locales/en-US/").unwrap();
        diag_handler.abort_if_errors();
        diag_handler.add_warn_diagnostic(Diagnostic::<DiagnosticStyle>::new());
        diag_handler.abort_if_errors();
        diag_handler.add_err_diagnostic(Diagnostic::<DiagnosticStyle>::new());

        let result = panic::catch_unwind(|| {
            diag_handler.abort_if_errors();
        });
        assert!(result.is_err());
    }
}
