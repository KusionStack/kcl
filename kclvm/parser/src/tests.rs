use std::{
    env,
    panic::{catch_unwind, set_hook},
};

use compiler_base_span::{FilePathMapping, SourceMap};
use kclvm_config::modfile::{get_vendor_home, KCLVM_VENDOR_HOME};

use crate::*;

use core::any::Any;

mod error_recovery;

#[macro_export]
macro_rules! parse_expr_snapshot {
    ($name:ident, $src:expr) => {
        #[test]
        fn $name() {
            insta::assert_snapshot!($crate::tests::parsing_expr_string($src));
        }
    };
}

#[macro_export]
macro_rules! parse_module_snapshot {
    ($name:ident, $src:expr) => {
        #[test]
        fn $name() {
            insta::assert_snapshot!($crate::tests::parsing_module_string($src));
        }
    };
}

pub(crate) fn parsing_expr_string(src: &str) -> String {
    let sm = SourceMap::new(FilePathMapping::empty());
    let sf = sm.new_source_file(PathBuf::from("").into(), src.to_string());
    let sess = &ParseSession::with_source_map(Arc::new(sm));

    match sf.src.as_ref() {
        Some(src_from_sf) => create_session_globals_then(|| {
            let stream = parse_token_streams(sess, src_from_sf.as_str(), new_byte_pos(0));
            let mut parser = Parser::new(sess, stream);
            let expr = parser.parse_expr();
            format!("{expr:#?}\n")
        }),
        None => "".to_string(),
    }
}

pub(crate) fn parsing_module_string(src: &str) -> String {
    let sm = SourceMap::new(FilePathMapping::empty());
    let sf = sm.new_source_file(PathBuf::from("").into(), src.to_string());
    let sess = &ParseSession::with_source_map(Arc::new(sm));

    match sf.src.as_ref() {
        Some(src_from_sf) => create_session_globals_then(|| {
            let stream = parse_token_streams(sess, src_from_sf.as_str(), new_byte_pos(0));
            let mut parser = Parser::new(sess, stream);
            let module = parser.parse_module();
            format!("{module:#?}\n")
        }),
        None => "".to_string(),
    }
}

pub fn check_result_panic_info(result: Result<(), Box<dyn Any + Send>>) {
    if let Err(e) = result {
        match e.downcast::<String>() {
            Ok(_v) => {
                let got = _v.to_string();
                let _u: PanicInfo = serde_json::from_str(&got).unwrap();
            }
            _ => unreachable!(),
        }
    };
}

const PARSE_EXPR_INVALID_TEST_CASES: &[&str] =
    &["fs1_i1re1~s", "fh==-h==-", "8_________i", "1MM", "0x00x"];

#[test]
pub fn test_parse_expr_invalid() {
    for case in PARSE_EXPR_INVALID_TEST_CASES {
        set_hook(Box::new(|_| {}));
        let result = catch_unwind(|| {
            parse_expr(case);
        });
        check_result_panic_info(result);
    }
}

const PARSE_FILE_INVALID_TEST_CASES: &[&str] = &[
    "a: int",                   // No initial value error
    "a -",                      // Invalid binary expression error
    "a?: int",                  // Invalid optional annotation error
    "a: () = 1",                // Type annotation error
    "if a not is not b: a = 1", // Logic operator error
    "if True:\n  a=1\n b=2",    // Indent error with recovery
    "a[1::::]",                 // List slice error
    "a[1 a]",                   // List index error
    "{a ++ 1}",                 // Config attribute operator error
    "func(a=1,b)",              // Call argument error
    "'${}'",                    // Empty string interpolation error
    "'${a: jso}'",              // Invalid string interpolation format spec error
];

#[test]
pub fn test_parse_file_invalid() {
    for case in PARSE_FILE_INVALID_TEST_CASES {
        let result = parse_file("test.k", Some((&case).to_string()));
        assert!(result.is_err(), "case: {case}, result {result:?}");
    }
}

pub fn test_vendor_home() {
    let vendor = &PathBuf::from(".")
        .join("testdata")
        .join("test_vendor")
        .canonicalize()
        .unwrap()
        .display()
        .to_string();
    env::set_var(KCLVM_VENDOR_HOME, vendor);
    assert_eq!(get_vendor_home(), vendor.to_string());
}

fn set_vendor_home() -> String {
    // set env vendor
    let vendor = &PathBuf::from(".")
        .join("testdata")
        .join("test_vendor")
        .canonicalize()
        .unwrap()
        .display()
        .to_string();
    env::set_var(KCLVM_VENDOR_HOME, vendor);
    debug_assert_eq!(get_vendor_home(), vendor.to_string());
    vendor.to_string()
}

#[test]
/// The testing will set environment variables,
/// so can not to execute test cases concurrently.
fn test_in_order() {
    test_import_vendor_without_vendor_home();
    println!("{:?} PASS", "test_import_vendor_without_vendor_home");
    test_import_vendor_without_kclmod();
    println!("{:?} PASS", "test_import_vendor_without_kclmod");
    test_import_vendor();
    println!("{:?} PASS", "test_import_vendor");
    test_import_vendor_with_same_internal_pkg();
    println!("{:?} PASS", "test_import_vendor_with_same_internal_pkg");
    test_import_vendor_without_kclmod_and_same_name();
    println!(
        "{:?} PASS",
        "test_import_vendor_without_kclmod_and_same_name"
    );
    test_vendor_home();
    println!("{:?} PASS", "test_vendor_home");
}

pub fn test_import_vendor() {
    let vendor = set_vendor_home();
    let sm = SourceMap::new(FilePathMapping::empty());
    let sess = Arc::new(ParseSession::with_source_map(Arc::new(sm)));

    let test_cases = vec![
        ("assign.k", vec!["__main__", "assign", "assign.assign"]),
        (
            "config_expr.k",
            vec!["__main__", "config_expr", "config_expr.config_expr_02"],
        ),
        (
            "nested_vendor.k",
            vec![
                "__main__",
                "nested_vendor",
                "nested_vendor.nested_vendor",
                "vendor_subpkg",
                "sub.sub1",
                "sub.sub2",
                "sub.sub",
                "sub",
            ],
        ),
        (
            "subpkg.k",
            vec![
                "__main__",
                "vendor_subpkg",
                "sub.sub1",
                "sub.sub",
                "sub.sub2",
                "sub",
            ],
        ),
    ];

    let dir = &PathBuf::from(".")
        .join("testdata")
        .join("import_vendor")
        .canonicalize()
        .unwrap();

    test_cases.into_iter().for_each(|(test_case_name, pkgs)| {
        let test_case_path = dir.join(test_case_name).display().to_string();
        let m = load_program(sess.clone(), &[&test_case_path], None).unwrap();
        assert_eq!(m.pkgs.len(), pkgs.len());
        m.pkgs.into_iter().for_each(|(name, modules)| {
            assert!(pkgs.contains(&name.as_str()));
            for pkg in pkgs.clone() {
                if name == pkg {
                    if name == "__main__" {
                        assert_eq!(modules.len(), 1);
                        assert_eq!(modules.get(0).unwrap().filename, test_case_path);
                    } else {
                        modules.into_iter().for_each(|module| {
                            assert!(module.filename.contains(&vendor));
                        });
                    }
                    break;
                }
            }
        });
    });
}

pub fn test_import_vendor_without_kclmod() {
    let vendor = set_vendor_home();
    let sm = SourceMap::new(FilePathMapping::empty());
    let sess = Arc::new(ParseSession::with_source_map(Arc::new(sm)));

    let test_cases = vec![("import_vendor.k", vec!["__main__", "assign.assign"])];

    let dir = &PathBuf::from(".")
        .join("testdata_without_kclmod")
        .canonicalize()
        .unwrap();

    test_cases.into_iter().for_each(|(test_case_name, pkgs)| {
        let test_case_path = dir.join(test_case_name).display().to_string();
        let m = load_program(sess.clone(), &[&test_case_path], None).unwrap();
        assert_eq!(m.pkgs.len(), pkgs.len());
        m.pkgs.into_iter().for_each(|(name, modules)| {
            assert!(pkgs.contains(&name.as_str()));
            for pkg in pkgs.clone() {
                if name == pkg {
                    if name == "__main__" {
                        assert_eq!(modules.len(), 1);
                        assert_eq!(modules.get(0).unwrap().filename, test_case_path);
                    } else {
                        modules.into_iter().for_each(|module| {
                            assert!(module.filename.contains(&vendor));
                        });
                    }
                    break;
                }
            }
        });
    });
}

pub fn test_import_vendor_without_vendor_home() {
    env::set_var(KCLVM_VENDOR_HOME, "");
    let sm = SourceMap::new(FilePathMapping::empty());
    let sess = Arc::new(ParseSession::with_source_map(Arc::new(sm)));
    let dir = &PathBuf::from(".")
        .join("testdata")
        .join("import_vendor")
        .canonicalize()
        .unwrap();
    let test_case_path = dir.join("assign.k").display().to_string();
    match load_program(sess, &[&test_case_path], None) {
        Ok(_) => {
            panic!("Unreachable code.")
        }
        Err(err) => {
            let result: PanicInfo = serde_json::from_str(&err).unwrap();
            assert_eq!(result.message, "pkgpath assign not found in the program");
        }
    }
}

fn test_import_vendor_with_same_internal_pkg() {
    set_vendor_home();
    let sm = SourceMap::new(FilePathMapping::empty());
    let sess = Arc::new(ParseSession::with_source_map(Arc::new(sm)));
    let dir = &PathBuf::from(".")
        .join("testdata")
        .join("import_vendor")
        .canonicalize()
        .unwrap();
    let test_case_path = dir.join("same_name.k").display().to_string();
    match load_program(sess, &[&test_case_path], None) {
        Ok(_) => {
            panic!("Unreachable code.")
        }
        Err(err) => {
            let result: PanicInfo = serde_json::from_str(&err).unwrap();
            assert_eq!(
                result.message,
                "the `same_vendor` is found multiple times in the current package and vendor package"
            );
        }
    }
}

fn test_import_vendor_without_kclmod_and_same_name() {
    set_vendor_home();
    let sm = SourceMap::new(FilePathMapping::empty());
    let sess = Arc::new(ParseSession::with_source_map(Arc::new(sm)));
    let dir = &PathBuf::from(".")
        .join("testdata_without_kclmod")
        .join("same_name")
        .canonicalize()
        .unwrap();
    let test_case_path = dir.join("assign.k").display().to_string();
    match load_program(sess, &[&test_case_path], None) {
        Ok(_) => {
            panic!("Unreachable code.")
        }
        Err(err) => {
            let result: PanicInfo = serde_json::from_str(&err).unwrap();
            assert_eq!(
                result.message,
                "the `assign` is found multiple times in the current package and vendor package"
            );
        }
    }
}
