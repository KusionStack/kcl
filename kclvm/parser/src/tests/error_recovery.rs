use crate::tests::parse_expr_snapshot;

parse_expr_snapshot! { string_literal_recovery_0, "'abc" }
parse_expr_snapshot! { string_literal_recovery_1, "r'abc" }
parse_expr_snapshot! { string_literal_recovery_2, "'''abc" }
parse_expr_snapshot! { string_literal_recovery_3, "r'''abc" }
parse_expr_snapshot! { string_literal_recovery_4, "r''abc'" }
parse_expr_snapshot! { string_literal_recovery_5, "'" }
parse_expr_snapshot! { string_literal_recovery_6, "'''" }
parse_expr_snapshot! { string_literal_recovery_7, "'\n" }
parse_expr_snapshot! { string_literal_recovery_8, "r'abc\n" }
parse_expr_snapshot! { number_literal_recovery_0, "00" }
parse_expr_snapshot! { number_literal_recovery_1, "00a" }
parse_expr_snapshot! { number_literal_recovery_2, "0x112.3" }
parse_expr_snapshot! { number_literal_recovery_3, "0o" }
parse_expr_snapshot! { number_literal_recovery_4, "0oA" }
parse_expr_snapshot! { number_literal_recovery_5, "0x" }
parse_expr_snapshot! { number_literal_recovery_6, "0xH" }
parse_expr_snapshot! { number_literal_recovery_7, "0e0" }
parse_expr_snapshot! { number_literal_recovery_8, "0b333" }
parse_expr_snapshot! { number_literal_recovery_9, "10KI" }
parse_expr_snapshot! { number_literal_recovery_10, "100mm" }
parse_expr_snapshot! { line_continue_recovery_0, "0x\\2\n12" }
parse_expr_snapshot! { line_continue_recovery_1, "'abc\\ \ndef" }
parse_expr_snapshot! { line_continue_recovery_2, r#"'a' + \
'b'
"# }
parse_expr_snapshot! { line_continue_recovery_3, r#"'a' + \1
'b'
"# }
parse_expr_snapshot! { paren_recovery_0, "(a" }
parse_expr_snapshot! { paren_recovery_1, "(a + 1" }
parse_expr_snapshot! { list_recovery_0, "[" }
parse_expr_snapshot! { list_recovery_1, "[0" }
parse_expr_snapshot! { list_recovery_2, "[0,1" }
parse_expr_snapshot! { list_recovery_3, "[[0,1]" }
parse_expr_snapshot! { list_recovery_4, "[[0,1" }
parse_expr_snapshot! { list_recovery_5, r#"[
    0,
    1
    "# }
parse_expr_snapshot! { config_recovery_0, "{" }
parse_expr_snapshot! { config_recovery_1, "{a = 1" }
parse_expr_snapshot! { config_recovery_2, "{a = 1, b = 2" }
parse_expr_snapshot! { config_recovery_3, "{a = {a = 1}" }
parse_expr_snapshot! { config_recovery_4, "{a = {a = 1" }
parse_expr_snapshot! { config_recovery_5, r#"{
    a = 1
    b = 2
    "# }
