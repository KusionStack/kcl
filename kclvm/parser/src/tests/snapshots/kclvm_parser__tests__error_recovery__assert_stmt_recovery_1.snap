---
source: parser/src/tests/error_recovery.rs
assertion_line: 161
expression: "crate::tests::parsing_module_string(r#\"assert a.\"#)"
---
Module {
    filename: "",
    pkg: "",
    doc: "",
    name: "",
    body: [
        Node {
            node: Assert(
                AssertStmt {
                    test: Node {
                        node: Identifier(
                            Identifier {
                                names: [
                                    "a",
                                    "",
                                ],
                                pos: [
                                    Node {
                                        node: (),
                                        filename: "",
                                        line: 1,
                                        column: 7,
                                        end_line: 1,
                                        end_column: 8,
                                    },
                                    Node {
                                        node: (),
                                        filename: "",
                                        line: 1,
                                        column: 9,
                                        end_line: 1,
                                        end_column: 9,
                                    },
                                ],
                                pkgpath: "",
                                ctx: Load,
                            },
                        ),
                        filename: "",
                        line: 1,
                        column: 7,
                        end_line: 1,
                        end_column: 9,
                    },
                    if_cond: None,
                    msg: None,
                },
            ),
            filename: "",
            line: 1,
            column: 0,
            end_line: 1,
            end_column: 9,
        },
    ],
    comments: [],
}

